src/var.c

```diff
@@ -53,8 +56,87 @@ struct fi_var_t {
 
 static struct fi_var_t *fi_vars = NULL;
 
+static int fi_var_get(const struct fi_provider *provider, const char *var_name,
+		char **value)
+{
+	struct fi_var_t *v;
+
+	// Check for bozo cases
+	if (var_name == NULL || value == NULL) {
+		FI_DBG(provider, FI_LOG_CORE,
+			"Failed to read %s variable: provider coding error\n",
+			var_name);
+		return -FI_EINVAL;
+	}
+
+	for (v = fi_vars; v; v = v->next) {
+		if (strcmp(v->provider->name, provider->name) == 0 &&
+			strcmp(v->var_name, var_name) == 0) {
+			*value = getenv(v->env_var_name);
+
+			return FI_SUCCESS;
+		}
+	}
+
+	FI_DBG(provider, FI_LOG_CORE,
+		"Failed to read %s variable: was not registered\n", var_name);
+	return -FI_ENOENT;
+}
+
+__attribute__((visibility ("default")))
+int DEFAULT_SYMVER_PRE(fi_getsettings)(struct fi_setting **vars, int *count) {
+
+	struct fi_setting *vhead = NULL;
+	struct fi_var_t *ptr;
+	int ret, len = 0, i = 0;
+	char *tmp;
+
+	// just get a count
+	for (ptr = fi_vars; ptr; ptr = ptr->next, ++len)
+		continue;
+
+	if (len == 0)
+		goto no_vars;
+
+	vhead = calloc(len, sizeof (*vhead));
+
+	if (!vhead)
+		return -FI_ENOMEM;
+
+	for (ptr = fi_vars; ptr; ptr = ptr->next, ++i) {
+		vhead[i].prov_name = strdup(ptr->provider->name);
+		vhead[i].var_name = strdup(ptr->var_name);
+		vhead[i].env_var_name = strdup(ptr->env_var_name);
+		vhead[i].help_string = strdup(ptr->help_string);
 
-int fi_var_register(const struct fi_provider *provider, const char *var_name,
+		/* ignore error for now */
+		ret = fi_var_get(ptr->provider, ptr->var_name, &tmp);
+
+		if (ret == FI_SUCCESS)
+			vhead[i].value = strdup(tmp);
```

Comment:
Add checks against strdup failing, so that coverity doesn't immediately complain after this is merged.